Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -477,57 +436,54 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
     ///
-    /// The resolved references are collected in `externallyResolvedLinks`.
+    /// The resolved references are collected in ``externallyResolvedLinks``.
     ///
     /// - Parameters:
     ///   - semanticObjects: A list of semantic objects to visit to collect links.
-    ///   - bundle: The documentation bundle to resolve links against.
-    private func preResolveExternalLinks(semanticObjects: [ReferencedSemanticObject], bundle: DocumentationBundle) throws {
+    ///   - localBundleID: The local bundle ID, used to identify and skip absolute fully qualified local links.
+    private func preResolveExternalLinks(semanticObjects: [ReferencedSemanticObject], localBundleID: BundleIdentifier) {
         // If there are no external resolvers added we will not resolve any links.
-        guard !externalReferenceResolvers.isEmpty else { return }
+        guard !externalDocumentationSources.isEmpty else { return }
         
-        let collectedExternalLinks: [ExternalLinkResult] = semanticObjects.concurrentPerform { object, results in
+        let collectedExternalLinks = Synchronized([String: Set<UnresolvedTopicReference>]())
+        semanticObjects.concurrentPerform { _, semantic in
             autoreleasepool {
                 // Walk the node and extract external link references.
-                var externalLinksCollector = ExternalReferenceWalker(bundle: bundle)
-                externalLinksCollector.visit(object.semantic)
+                var externalLinksCollector = ExternalReferenceWalker(localBundleID: localBundleID)
+                externalLinksCollector.visit(semantic)
+
+                // Avoid any synchronization overhead if there are no references to add.
+                guard !externalLinksCollector.collectedExternalReferences.isEmpty else { return }
                 
                 // Add the link pairs to `collectedExternalLinks`.
-                results.append(contentsOf:
-                    externalLinksCollector.collectedExternalReferences.map { unresolved -> ExternalLinkResult in
-                        return ExternalLinkResult(unresolved: unresolved, targetLanguage: object.reference.sourceLanguage)
+                collectedExternalLinks.sync {
+                    for (bundleID, collectedLinks) in externalLinksCollector.collectedExternalReferences {
+                        $0[bundleID, default: []].formUnion(collectedLinks)
                     }
-                )
+                }
             }
         }
         
-        try Set(collectedExternalLinks).compactMap { externalLink -> (url: ValidatedURL, resolved: TopicReferenceResolutionResult)? in
-            guard let referenceBundleIdentifier = externalLink.unresolved.topicURL.components.host else {
-                assertionFailure("Should not hit this code path, url is verified to have an external bundle id.")
-                return nil
+        for (bundleID, collectedLinks) in collectedExternalLinks.sync({ $0 }) {
```

Comment:
What does the `.sync({ $0 })` achieve here?

Reply:
It accesses the wrapped value. The `collectedExternalLinks` variable is using our `Synchronized` wrapper, so all the reads and write need to be synchronized. Here `.sync({ $0 })` is implicitly returning the wrapped value meaning that it's the same as `.sync({ wrappedValue in return wrappedValue })`